File: round-table/round-table.py
from sys import maxsize, stdin
from math import ceil


def get_list(): return list(map(int, stdin.readline().strip().split()))
def get_ints(): return list(map(int, stdin.readline().strip().split()))


def customBinSearch(list, val):
    low = 0
    high = len(list) - 1
    mid = (high + low) // 2
    while low < high:
        if list[mid] > val:
            high = mid
        else:
            low = mid + 1
        mid = (high + low) // 2
    if list[mid] > val:
        return list[mid - 1 if mid - 1 >= 0 else 0], list[mid]
    else:
        return list[mid], list[mid + 1 if mid + 1 < len(list) else len(list) - 1]


if __name__ == "__main__":
    N, Q = get_ints()

    mapIndexes = {}
    arr = get_list()
    def addN(x): return x + N
    def subN(x): return x - N
    for i in range(N):
        e = arr[i]
        if e not in mapIndexes:
            mapIndexes[e] = []
        if i not in mapIndexes[e]:
            mapIndexes[e].append(i)

    for _ in range(Q):
        min_dist = maxsize
        x, y = get_ints()
        extendedIndexes = list(
            map(subN, mapIndexes[y])) + mapIndexes[y] + list(map(addN, mapIndexes[y]))
        for i in range(len(mapIndexes[x])):
            start = mapIndexes[x][i]
            left, right = customBinSearch(extendedIndexes, start)
            distance1 = ceil((abs(left - start) - 1) / 2)
            distance2 = ceil((abs(right - start) - 1) / 2)
            min_dist = min(min_dist, distance1, distance2)

        print(min_dist)

File: round-table/round-table.js
const readline = require("readline");
const rl = readline.createInterface({ input: process.stdin });
let inputs = [];

rl.on("line", function (line) {
  line
    .trim()
    .split(/[ ,]+/)
    .forEach((e) => inputs.push(e));
});

rl.on("close", function () {
  main();
});

function main() {
  let N = inputs.splice(0, 1);
  let Q = inputs.splice(0, 1);
  const arr = inputs.splice(0, N);
  let mapIndexes = {};
  for (let i = 0; i < N; i++) {
    const e = +arr[i];
    if (!(e in mapIndexes)) mapIndexes[e] = [];
    if (!mapIndexes[e].includes(i)) mapIndexes[e].push(i);
  }

  let min, x, y, start, searchRes, left, right;

  while (Q--) {
    min = Number.MAX_SAFE_INTEGER;
    x = inputs.splice(0, 1);
    y = inputs.splice(0, 1);
    extendedIndexes = mapIndexes[y]
      .map((x) => x - N)
      .concat(mapIndexes[y])
      .concat(mapIndexes[y].map((x) => x + +N));
    for (let i = 0; i < mapIndexes[x].length; i++) {
      start = mapIndexes[x][i];

      searchRes = customBinSearch(extendedIndexes, start);
      left = searchRes[0];
      right = searchRes[1];
      distance1 = Math.ceil((Math.abs(left - start) - 1) / 2);
      distance2 = Math.ceil((Math.abs(right - start) - 1) / 2);
      min = Math.min(distance1, distance2, min);
    }
    console.log(min.toFixed(0).replace('-0', '0'));
  }
}

function customBinSearch(list, val) {
  let low = 0;
  let high = list.length - 1;
  let mid = Math.floor((high + low) / 2);
  while (low < high) {
    if (list[mid] > val) high = mid;
    else low = mid + 1;
    mid = Math.floor((high + low) / 2);
  }
  return list[mid] > val
    ? [list[mid - 1 >= 0 ? mid - 1 : 0], list[mid]]
    : [list[mid], list[mid + 1 < list.length ? mid + 1 : list.length - 1]];
}
